aeron/aeron.go

```diff
@@ -110,48 +114,91 @@ func (aeron *Aeron) Close() error {
 	return err
 }
 
-// AddSubscription will add a new subscription to the driver.
+// AddSubscription will add a new subscription to the driver and wait until it is ready.
+func (aeron *Aeron) AddSubscription(channel string, streamID int32) (*Subscription, error) {
+	registrationID, err := aeron.conductor.AddSubscription(channel, streamID)
+	if err != nil {
+		return nil, err
+	}
+	for {
+		subscription, err := aeron.conductor.FindSubscription(registrationID)
+		// Retry temporary error.  Any other success or error gets returned.
+		if !errors.Is(err, TemporaryError) {
+			return subscription, err
+		}
+		aeron.context.idleStrategy.Idle(0)
+	}
+}
+
+// AddSubscriptionDeprecated will add a new subscription to the driver.
 // Returns a channel, which can be used for either blocking or non-blocking want for media driver confirmation
-func (aeron *Aeron) AddSubscription(channel string, streamID int32) chan *Subscription {
+func (aeron *Aeron) AddSubscriptionDeprecated(channel string, streamID int32) chan *Subscription {
 	ch := make(chan *Subscription, 1)
-
-	regID := aeron.conductor.AddSubscription(channel, streamID)
+	registrationID, err := aeron.conductor.AddSubscription(channel, streamID)
+	if err != nil {
+		// Preserve the legacy functionality.  The original AddSubscription would result in the ClientConductor calling
+		// onError on this, as well as subsequently from the FindSubscription call below.
+		aeron.conductor.onError(err)
+	}
 	go func() {
 		for {
 			// This is a quirk of retrofitting proper error handling onto an API that does not return an error value.
 			// FindSubscription can return one of 3 states:
 			// nil, nil: Still waiting, keep polling
 			// subscription, nil: Success, return subscription
 			// nil, error: Permanent failure, cc.onError() has already been called, exit gracefully
-			subscription, err := aeron.conductor.FindSubscription(regID)
+			subscription, err := aeron.conductor.FindSubscription(registrationID)
```

Comment:
this is missing the temporary error logic (I think)

```
		subscription, err := aeron.conductor.FindSubscription(registrationID)
		// Retry temporary error.  Any other success or error gets returned.
		if !errors.Is(err, TemporaryError) {
			return subscription, err
		}
```

Reply:
Probably the same for AddPublicationDeprecated

Reply:
This does the trick 

```
			if publication != nil || (err != nil && !errors.Is(err, TemporaryError)) {
```

Reply:
Nice catch, done